content/courses/advanced-ada/chapters/exceptions.rst

```diff
@@ -247,9 +247,172 @@ Out and Uninitialized
     This section was originally written by Robert Dewar and published as
     `Gem #150: Out and Uninitialized <https://www.adacore.com/gems/gem-150out-and-uninitialized>`_
 
-.. todo::
+Perhaps surprisingly, the Ada standard indicates cases where objects passed to
+:ada:`out` and :ada:`in out` parameters might not be updated when a procedure
+terminates due to an exception. Let's take an example:
 
-    Complete section!
+.. code:: ada run_button project=Courses.Advanced_Ada.Exceptions.Out_Uninitialized_1
+
+    with Ada.Text_IO;  use Ada.Text_IO;
+    procedure Show_Out_Uninitialized is
+
+       procedure Local (A : in out Integer; Error : Boolean) is
+       begin
+          A := 1;
+
+          if Error then
+             raise Program_Error;
+          end if;
+       end Local;
+
+       B : Integer := 0;
+
+    begin
+       Local (B, Error => True);
+    exception
+       when Program_Error =>
+          Put_Line ("Value for B is" & Integer'Image (B));  --  "0"
+    end Show_Out_Uninitialized;
+
+This program outputs a value of 0 for :ada:`B`, whereas the code indicates that
+:ada:`A` is assigned before raising the exception, and so the reader might
+expect :ada:`B` to also be updated.
+
+The catch, though, is that a compiler must by default pass objects of
+elementary types (scalars and access types) by copy and might choose to do so
+for other types (records, for example), including when passing :ada:`out` and
+:ada:`in out` parameters. So what happens is that while the formal parameter
+:ada:`A` is properly initialized, the exception is raised before the new value
+of :ada:`A` has been copied back into :ada:`B` (the copy will only happen on a
+normal return).
+
+.. admonition:: In the GNAT toolchain
+
+    In general, any code that reads the actual object passed to an :ada:`out` or
+    :ada:`in out` parameter after an exception is suspect and should be avoided.
+    GNAT has useful warnings here, so that if we simplify the above code to:
+
+    .. code:: ada run_button project=Courses.Advanced_Ada.Exceptions.Out_Uninitialized_2
+
+        with Ada.Text_IO;  use Ada.Text_IO;
+
+        procedure Show_Out_Uninitialized_Warnings is
+
+            procedure Local (A : in out Integer) is
+            begin
+               A := 1;
+               raise Program_Error;
+            end Local;
+
+           B : Integer := 0;
+
+        begin
+           Local (B);
+        exception
+           when others =>
+              Put_Line ("Value for B is" & Integer'Image (B));
+        end Show_Out_Uninitialized_Warnings;
+
+    We now get a compilation warning that the pass-by-copy formal may have no
+    effect.
+
+    Of course, GNAT is not able to point out all such errors (see first example
+    above), which in general would require full flow analysis.
+
+The behavior is different when using parameter types that the standard mandates
+passing by reference, such as tagged types for instance. So the following code
```

Comment:
"The behavior is different when using parameter types that the standard mandates passing by reference, such as tagged types for instance." is a little awkward (but not much). 
Maybe:
"The behavior is different when using parameter types that the standard mandates be passed by reference, such as tagged types for instance."
Or:
"The behavior is different when using parameter types that are passed by reference, such as tagged types for instance."